### ISSUE TITLE
Expand codepipeline.Pipeline public API (and others) for better post-instantiation customizability

### ISSUE THREAD

**Issue by michaelmoussa** [aws/aws-cdk#4878]
Presently, once a pipeline object has been created, the ability to make modifications to its stages is _very_ limited. If I'm not mistaken, one is only able to add new stages, which will end up at the end of the pipeline unless they already have a reference available to an existing stage for `rightBefore` / `justAfter` placement.

I would like for `CodePipeline.Pipeline` to have a public API that better exposes its `stages` and their underlying `actions` for customizability after they've been instantiated. 

### Use Case

Suppose I have a custom, internal L3 construct for my organization called `DefaultPipeline` which teams `npm install` from our private repository. The construct extends `codepipeline.Pipeline` and its purpose is to provide all teams with a consistent base pipeline that grabs source from GitHub, performs build actions to run `cdk synth` on the CDK infra scripts and build the application code itself, sends a notification for Manual Approval, and then finally has a deploy step that deploys the CDK infra changes and the application.

It's quite simple to use as well. All a team needs to do is:

```typescript
import { DefaultPipeline } from "my-organization/default-pipeline;"

export class MyStack extends Stack {
    constructor(...) {
        super(...)

        new DefaultPipeline(this, "MyPipeline", { ... repo info goes here ... })
    }
}
```

That's it! The `DefaultPipeline` construct takes care of the rest.

This hypothetical construct works perfectly well for 99.999% of instances that my hypothetical organization has used it, but one team needs to make 3 modifications:

1. They want to add a stage between **Build** and **Deploy** that executes a **Lambda** function.

2. They need to add an environment variable called `FOO` with the value `BAR` to the **BuildApplication** action of the **Build** stage.

3. They want to remove the **Manual Approval** stage entirely.

Since this is the only team that needs these modifications, it's less-than-ideal to require the `DefaultPipeline` maintainers to provide for this customizability. Further, since `codepipeline.Pipeline` does not expose very much of `stages` at all, the `DefaultPipeline` maintainers would need to provide for the required customizability *in the constructor* because once the Pipeline object has been created, it's already too late. This will end up bloating the constructor and making things difficult to maintain.

The other option is for the team to either fork the `DefaultPipeline` L3 construct in order to make their modifications, or copy-and-paste it entirely. This is also less-than-ideal because it will potentially make incorporating backwards-compatible updates to the upstream `DefaultPipeline` more difficult or impossible. Copy-and-paste is definitely not my first approach to code reuse. :)

The only other approach I can think of is to rely on the `.node.defaultChild.<...>` escape hatch and apply the tweaks to the underlying L1 elements. This will work, but escape hatches are meant as more of a temporary "oh, there's a missing feature or you discovered a bug? do what you gotta do to work around it until it's implemented/fixed" rather than a permanent "I don't know if CDK will ever support this, but hopefully a future release won't break this hack of mine!"

### Proposed Solution

I'd like for `codepipeline.Pipeline` to do a better job of providing for these post-instantiation customizations. I'm focusing specifically on `codepipeline.Pipeline` in this issue because it is what prompted me to create it, but the spirit of this feature request applies to every core L2 construct where the public API is fairly limited.

Here's how I'd envision some public API changes making the three use cases I mentioned easily solvable without having to modify the L3 construct source code:

1. They want to add a stage between **Build** and **Deploy** that executes a **Lambda** function.

```typescript
pipeline.addStage({
    stageName: "...",
    actions: [new LambdaInvokeAction(...)],
    placement: {justAfter: pipeline.getStage("NameOfBuildStage")}
})
```

`codepipeline.Pipeline` currently lets you add a new stage, but unless you have a reference to an existing stage to insert it before/after, `addStage(...)` will add it to the end of the pipeline. I don't believe there's currently a way to get a reference to an existing stage within a `Pipeline` object after it has been instantiated.

2. They need to add an environment variable called `FOO` with the value `BAR` to the **BuildApplication** action of the **Build** stage.

```typescript
const buildAction = pipeline.getStage("NameOfBuildStage").getAction("NameOfCodeBuildAction")
buildAction.setEnvironmentVariable(
    "FOO",
    {type: BuildEnvironmentVariableType.PLAINTEXT, value: "BAR"}
)
```

Assuming `.getStage(...)` and `.getAction(...)` return mutable references, we're done. If they return copies, a `.replaceStage(...)` or `.replaceAction(...)` mechanism would be needed.

3. They want to remove the **Manual Approval** stage entirely.

```typescript
pipeline.removeStage(pipeline.getStage("NameOfApprovalStage"));
```

Too easy. What if **Manual Approval** was an action at the end of the **Build** stage instead? Then we could just do this:

```typescript
const buildStage = pipeline.getStage("NameOfBuildStage");
buildStage.removeAction(buildStage.getAction("NameOfActionToRemove"));
```

### Other

There are a number of different ways this could be done for `codepipeline.Pipeline` specifically (`get/set`, `add/remove`, `replace`, `unset` function prefixes just to name a few), but if this is something the project would be interested in considering, I think it'd be a good idea to standardize what the API would look like and make it a point to ensure that any similar customizability that is added to other L2 constructs is implemented in a consistent way.

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

**NOTE:** w.r.t. breaking changes, I believe this can be done in such a way that preserves BC in CDK 1.x. If we want to expand this functionality to other L2 constructs _or_ implement it in such a way that would end up breaking BC, we could deprecate whatever would be affected and apply the BC break in a major release.

---

This is a :rocket: Feature Request

**Comment by mbonig** [comment]
👍 the richer API on CodePipeline.

**Comment by skinny85** [comment]
Hey @michaelmoussa ,

thanks for opening the issue. Here are my thoughts on each of your suggestions.

### 1. Pipeline.getStage(stageName: string) method

I'm actually not in love with this method. This means you have to hard-code the names of the stages in the consumer.

**Alternative**: how about surfacing the stages as part of the `DefaultPipeline` construct?

```ts
const pipeline = new DefaultPipeline(this, "MyPipeline", { ... repo info goes here ... });
pipeline.addStage({
    stageName: "...",
    actions: [new LambdaInvokeAction(...)],
    placement: {
      justAfter: pipeline.buildStage,
  },
});
```

### 2. Getting and modifying the actions already in the pipeline

The problem here is that this will not compile. Let's take your example:

```ts
const buildAction = pipeline.getStage("NameOfBuildStage").getAction("NameOfCodeBuildAction")
buildAction.setEnvironmentVariable(
    "FOO",
    {type: BuildEnvironmentVariableType.PLAINTEXT, value: "BAR"}
)
```

The type returned from `Stage.getAction()` would be `IAction`. `IAction` cannot have a `setEnvironmentVariable` method, because it's specific to `CodeBuildAction`. I don't want to force consumers to use casts in order to use the API, and we can't have generics in a JSII-compatible library.

**Alternative**: why don't you surface this an option when creating the `DefaultPipeline`?

```ts
new DefaultPipeline(this, "MyPipeline", {
  // other options...
  buildEnvironmentVariables: {
    FOO: {
      value: "BAR", // PLAINTEXT is the default type
    },
  },
});
```

### 3. Removing Stages and/or Actions

This is actually very problematic. The issue is that adding actions to a pipeline mutates a lot of resources. For example, in the (seemingly simple) case of the `ManualApprovalAction`, it:

1. Generates a new IAM `Role`.
2. If any notification emails have been provided, it creates an SNS Topic.
3. It grants the Role created in 1. permissions to the Topic created in 2.
4. It grants the pipeline's execution role `AssumeRole` permissions to the Role created in 1.

How would you unwind all of these mutations if the consumer goes `buildStage.removeAction('NameOfActionToRemove')`? Would we leave a bunch of dangling resources and permissions in the template, that are not used anywhere?

**Alternative**: why don't you make it configurable in `DefaultPipeline`?

```ts
new DefaultPipeline(this, "MyPipeline", {
  haveManualApproval: false, // default: true
});
```

I agree we want some customizations to be possible, but I think we need to plan them carefully, and understand the consequences.

**Comment by michaelmoussa** [comment]
Thanks for taking a look so quickly. I'll respond to your individual suggestions, but note that the examples I provided were just examples to get the discussion going, and not necessarily the best way to solve it.

> ### 1. Pipeline.getStage(stageName: string) method
> I'm actually not in love with this method. This means you have to hard-code the names of the stages in the consumer.

Right. It's not ideal, but I don't think it's much different from hardcoding other things one would normally hardcode in a construct, is it? Take the `cdk init sample-app --language=typescript` example that creates an SQS Queue with a hardcoded `visibilityTimeout` of `300` seconds. Hardcoding `300` seconds there could be justified if the value is going to be the same regardless of the environment it's being deployed to and if the value isn't expected to change.

In the case of my example, if the name of the build stage is `Build`, it's probably going to be `Build` always. If my dependency on the `DefaultPipeline` L3 construct is version-locked, then it's never going to change unless I update that dependency. If I *do* update the dependency and it ends up breaking my modification, it'd be caught by an automated test or local dev build.

It could *probably* be a little less smelly by using a constant in the consumer class along the lines of `DEFAULT_PIPELINE_BUILD_STAGE_NAME` and referencing that, or passing a property to the consumer stack that uses the `DefaultPipeline` L3 construct so that it can be updated without having to change the class.
 
> **Alternative**: how about surfacing the stages as part of the `DefaultPipeline` construct?

Yes, that would absolutely work, but I'm thinking beyond just stages within just Pipelines. What about every other potential L3 construct (both internal managed by another team or external open source managed by a third party)? The only thing the consumer in this case would be missing is the reference to the build stage object for the `placement.justAfter` setting.

I think it's a much bigger burden on someone maintaining an L3 Pipeline construct to be expected to surface the individual stages as they get created than it would be for the CDK to add a `getStage(<name>)` or even `getStages()` accessor to let consumers find the reference themselves.

> ### 2. Getting and modifying the actions already in the pipeline
> The problem here is that this will not compile. Let's take your example:
> 
> ```ts
> const buildAction = pipeline.getStage("NameOfBuildStage").getAction("NameOfCodeBuildAction")
> buildAction.setEnvironmentVariable(
>     "FOO",
>     {type: BuildEnvironmentVariableType.PLAINTEXT, value: "BAR"}
> )
> ```
> 
> The type returned from `Stage.getAction()` would be `IAction`. `IAction` cannot have a `setEnvironmentVariable` method, because it's specific to `CodeBuildAction`. I don't want to force consumers to use casts in order to use the API, and we can't have generics in a JSII-compatible library.

How about using a type guard instead of casting?

```typescript
if (buildAction instanceof CodeBuildAction) {
    buildAction.setEnvironmentVariable(...);
}
```

I wouldn't want to force casting either, but using a type guard here feels reasonable, especially if it's to facilitate a customization that would otherwise be a lot more work.
 
> **Alternative**: why don't you surface this an option when creating the `DefaultPipeline`?

Same as my response in (1). There would be much more work required by the `DefaultPipeline` L3 construct maintainer to anticipate every conceivable customization a consumer might want to do in comparison to some variation of what I'm proposing in this issue.

> ### 3. Removing Stages and/or Actions
> How would you unwind all of these mutations if the consumer goes `buildStage.removeAction('NameOfActionToRemove')`? Would we leave a bunch of dangling resources and permissions in the template, that are not used anywhere?

Great catch! I haven't dug deeply enough into CDK's internals to come up with a specific solution for this yet. CDK knows to delete stage- or action-related resources if I remove a particular stage or action from the actual source code (or rather, it knows _not_ to include those no-longer-present additional resources in the new `synth` output), so I think the answer lies in seeing if there's a reliable way to do the inverse.
 
> **Alternative**: why don't you make it configurable in `DefaultPipeline`?

Same as response in (1) and (2).

> I agree we want some customizations to be possible, but I think we need to plan them carefully, and understand the consequences.

I completely agree. I don't have any particular attachment to the specifics of what I proposed above and am happy to see if we can find a better approach. I just think a lot more can be done on the CDK side before leaving the rest as an implementation detail for the hypothetical construct maintainer.
